src/basic/json.c

```diff
@@ -0,0 +1,3318 @@
+                JsonVariant *reference;
+
+                /* Strings are placed immediately after the structure. Note that when this is a JsonVariant embedded
+                 * into an array we might encode strings up to INLINE_STRING_LENGTH characters directly inside the
+                 * element, while longer strings are stored as references. When this object is not embedded into an
+                 * array, but stand-alone we allocate the right size for the whole structure, i.e. the array might be
+                 * much larger than INLINE_STRING_LENGTH. */
+                char string[0];
+        };
+};
+
+/* Inside string arrays we have a series of JasonVariant structures one after the other. In this case, strings longer
+ * than INLINE_STRING_MAX are stored as references, and all shorter ones inline. (This means — on x86-64 — strings up
+ * to 15 chars are stored within the array elements, and all others in separate allocations) */
+#define INLINE_STRING_MAX (sizeof(JsonVariant) - offsetof(JsonVariant, string) - 1U)
+
+/* Let's make sure this structure isn't increased in size accidentally. This check is only for our most relevant arch
+ * (x86-64). */
+#ifdef __x86_64__
+assert_cc(sizeof(JsonVariant) == 48U);
+assert_cc(INLINE_STRING_MAX == 15U);
+#endif
+
+static JsonSource* json_source_new(const char *name) {
+        JsonSource *s;
+
+        assert(name);
+
+        s = malloc(offsetof(JsonSource, name) + strlen(name) + 1);
+        if (!s)
+                return NULL;
+
+        s->n_ref = 1;
+        s->max_line = 0;
+        s->max_column = 0;
+        strcpy(s->name, name);
+
+        return s;
+}
+
+static JsonSource* json_source_ref(JsonSource *s) {
+        if (!s)
+                return NULL;
+
+        assert(s->n_ref > 0);
+        s->n_ref++;
+
+        return s;
+}
+
+static JsonSource *json_source_unref(JsonSource *s) {
+        if (!s)
+                return NULL;
+
+        assert(s->n_ref > 0);
+        s->n_ref--;
+
+        if (s->n_ref > 0)
+                return NULL;
+
+        return mfree(s);
+}
+
+static bool json_source_equal(JsonSource *a, JsonSource *b) {
+        if (a == b)
+                return true;
+
+        if (!a || !b)
+                return false;
+
+        return streq(a->name, b->name);
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(JsonSource*, json_source_unref);
+
+static bool json_variant_is_magic(const JsonVariant *v) {
+        return v == JSON_VARIANT_MAGIC_TRUE ||
+                v == JSON_VARIANT_MAGIC_FALSE ||
+                v == JSON_VARIANT_MAGIC_NULL ||
+                v == JSON_VARIANT_MAGIC_ZERO_INTEGER ||
+                v == JSON_VARIANT_MAGIC_ZERO_UNSIGNED ||
+                v == JSON_VARIANT_MAGIC_ZERO_REAL ||
+                v == JSON_VARIANT_MAGIC_EMPTY_STRING ||
+                v == JSON_VARIANT_MAGIC_EMPTY_ARRAY ||
+                v == JSON_VARIANT_MAGIC_EMPTY_OBJECT;
```

Comment:
`IN_SET()`?

Reply:
can't use that unfortunately, v is a ptr, and IN_SET (which is built around a switch statement) doesn't support pointers... :-(